plugins/osdn/node_firewall.go

```diff
@@ -0,0 +1,102 @@
+package osdn
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	utildbus "k8s.io/kubernetes/pkg/util/dbus"
+	kexec "k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/iptables"
+	utilwait "k8s.io/kubernetes/pkg/util/wait"
+)
+
+type FirewallRule struct {
+	table string
+	chain string
+	args  []string
+}
+
+type Firewall struct {
+	ipt                iptables.Interface
+	clusterNetworkCIDR string
+	syncPeriod         time.Duration
+
+	mu sync.Mutex // Protects concurrent access to syncIPTableRules()
+}
+
+func NewNodeFirewall(clusterNetworkCIDR string, syncPeriod time.Duration) *Firewall {
+	return &Firewall{
+		ipt:                iptables.New(kexec.New(), utildbus.New(), iptables.ProtocolIpv4),
+		clusterNetworkCIDR: clusterNetworkCIDR,
+		syncPeriod:         syncPeriod,
+	}
+}
+
+func (f *Firewall) Setup() error {
+	err := f.syncIPTableRules()
+	if err != nil {
+		return err
+	}
+
+	// If firewalld is running, reload will call this method
+	f.ipt.AddReloadFunc(func() {
+		err := f.syncIPTableRules()
+		if err != nil {
+			glog.Errorf("Reloading openshift iptables failed: %v", err)
+		}
+	})
+
+	go utilwait.Forever(f.syncLoop, 0)
+	return nil
+}
+
+// syncLoop periodically calls syncIPTableRules().
+// This is expected to run as a go routine or as the main loop. It does not return.
+func (f *Firewall) syncLoop() {
+	t := time.NewTicker(f.syncPeriod)
+	defer t.Stop()
+	for {
+		<-t.C
+		glog.V(6).Infof("Periodic openshift iptables sync")
+		err := f.syncIPTableRules()
+		if err != nil {
+			glog.Errorf("Syncing openshift iptables failed: %v", err)
+		}
+	}
+}
+
+// syncIPTableRules syncs the cluster network cidr iptables rules.
+// Called from SyncLoop() or firwalld reload()
+func (f *Firewall) syncIPTableRules() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	start := time.Now()
+	defer func() {
+		glog.V(4).Infof("syncIPTableRules took %v", time.Since(start))
+	}()
+	glog.V(3).Infof("Syncing openshift iptables rules")
+
+	rules := f.getStaticNodeFirewallRules()
+	for _, rule := range rules {
+		_, err := f.ipt.EnsureRule(iptables.Prepend, iptables.Table(rule.table), iptables.Chain(rule.chain), rule.args...)
+		if err != nil {
+			return fmt.Errorf("Failed to ensure rule %v exists: %v", rule, err)
+		}
+	}
+	return nil
+}
+
+// Get openshift iptables rules
+func (f *Firewall) getStaticNodeFirewallRules() []FirewallRule {
+	return []FirewallRule{
+		{"nat", "POSTROUTING", []string{"-s", f.clusterNetworkCIDR, "!", "-d", f.clusterNetworkCIDR, "-j", "MASQUERADE"}},
+		{"filter", "INPUT", []string{"-p", "udp", "-m", "multiport", "--dports", "4789", "-m", "comment", "--comment", "001 vxlan incoming", "-j", "ACCEPT"}},
+		{"filter", "INPUT", []string{"-i", "tun0", "-m", "comment", "--comment", "traffic from docker for internet", "-j", "ACCEPT"}},
```

Comment:
Can you arrange for `ovs.TUN` to get passed down to here so that we don't have to hardcode `"tun0"`? (If it's not easy/clean then never mind, it will probably be easier after #305.)